seccomp GA - Add new seccomp fields and update kubelet to use them

pjbgf opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds the SeccompProfile field to both Pod and Container. It also implements the kubelet enforcement priority defined at the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md).

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of #91286

**Special notes for your reviewer**:
A separate PR will be created to cover annotation to field conversion, and its test cases.
The API changes made are based off the merged [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#pod-api).

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Seccomp support has graduated to GA. A new `seccompProfile` field is added to pod and container securityContext objects. Support for `seccomp.security.alpha.kubernetes.io/pod` and `container.seccomp.security.alpha.kubernetes.io/...` annotations is deprecated, and will be removed in v1.22.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig node
/area security

pjbgf commented:
/retest

pjbgf commented:
/retest

pjbgf commented:
The API changes are based off the already reviewed [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#pod-api).

/label api-review

pjbgf commented:
/cc @saschagrunert @hasheddan @evrardjp

saschagrunert commented on pkg/apis/core/types.go:
```diff
@@ -5070,6 +5098,11 @@ type SecurityContext struct {
 	// readonly paths and masked paths.
 	// +optional
 	ProcMount *ProcMountType
+	// The seccomp options to use by this container. If seccomp options are
+	// provided at both the pod & container level, the container options
+	// override the pod options.
+	// +optional
+	SeccompProfile *SeccompProfile
```
Maybe you know this out of your mind, so allow me one question:

Does it work if the sandbox profile has higher restrictions than the container? For example:
1. Pod sandbox runs on a seccomp profile which forbids `mkdir(2)`
2. Container now runs on a different profile allowing `mkdir(2)`
3. Executing `mkdir` will work inside the container

saschagrunert replied:
Verified that it works as intended by my own :heavy_check_mark:

saschagrunert commented on staging/src/k8s.io/api/core/v1/generated.proto:
```diff
@@ -4512,6 +4530,12 @@ message SecurityContext {
+  // provided at both the pod & container level, the container options
+  // override the pod options.
+  // +optional
+  optional SeccompProfile seccompProfile = 11;
```
Why is it field `11` and not `10`?

pjbgf replied:
This was automatically updated with `make update`. Going through the file there is already a number `10`:

https://github.com/kubernetes/kubernetes/blob/6115f15de19307871ff09ee950939f426b6ab56d/staging/src/k8s.io/api/core/v1/generated.proto#L4488

saschagrunert reviewed (commented):
Generally LGTM, I've just got a few nits/suggestions :point_down:

saschagrunert commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -202,8 +202,37 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus) *kubecontainer.Runtim
+		}
+		if scmp.Type == v1.SeccompProfileLocalhost && scmp.LocalhostProfile != nil {
+			fname := filepath.Join(m.seccompProfileRoot, *scmp.LocalhostProfile)
+			return "localhost/" + fname
```
I think we should make `localhost/` a constant somewhere. We have the same for AppArmor and we will need it in various places.

pjbgf replied:
Agreed, just updated accordingly.

pjbgf commented on pkg/apis/core/types.go:
```diff
@@ -2892,8 +2892,36 @@ type PodSecurityContext struct {
+	SeccompProfileUnconfined SeccompProfileType = "Unconfined"
+	// SeccompProfileTypeDefault represents the default container runtime seccomp profile.
+	SeccompProfileTypeDefault SeccompProfileType = "RuntimeDefault"
+	// SeccompProfileLocalhost represents custom made profiles stored on the node's disk.
```
I added that and also the mentioned that the flag is now deprecated, and what the location will be once the flag support is removed.

saschagrunert reviewed (approved):
/lgtm

hasheddan reviewed (approved):
/lgtm

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -2892,8 +2892,36 @@ type PodSecurityContext struct {
+
+const (
+	// SeccompProfileUnconfined is when no seccomp profile is applied (A.K.A. unconfined).
+	SeccompProfileUnconfined SeccompProfileType = "Unconfined"
```
stay consistent with SeccompProfileTypeDefault, so SeccompProfileTypeUnconfined

liggitt replied:
```suggestion
	SeccompProfileTypeUnconfined SeccompProfileType = "Unconfined"
```

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -2892,8 +2892,36 @@ type PodSecurityContext struct {
+	// SeccompProfileUnconfined is when no seccomp profile is applied (A.K.A. unconfined).
+	SeccompProfileUnconfined SeccompProfileType = "Unconfined"
+	// SeccompProfileTypeDefault represents the default container runtime seccomp profile.
+	SeccompProfileTypeDefault SeccompProfileType = "RuntimeDefault"
```
```suggestion
	SeccompProfileTypeRuntimeDefault SeccompProfileType = "RuntimeDefault"
```

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -2892,8 +2892,36 @@ type PodSecurityContext struct {
+	// SeccompProfileTypeDefault represents the default container runtime seccomp profile.
+	SeccompProfileTypeDefault SeccompProfileType = "RuntimeDefault"
+	// SeccompProfileLocalhost represents custom made profiles stored on the node's disk.
+	SeccompProfileLocalhost SeccompProfileType = "Localhost"
```
```suggestion
	SeccompProfileTypeLocalhost SeccompProfileType = "Localhost"
```

liggitt commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -286,6 +286,22 @@ func Convert_core_PodStatus_To_v1_PodStatus(in *core.PodStatus, out *v1.PodStatu
 	return nil
 }
 
+func Convert_core_SecurityContext_To_v1_SecurityContext(in *core.SecurityContext, out *v1.SecurityContext, s conversion.Scope) error {
+	if err := autoConvert_core_SecurityContext_To_v1_SecurityContext(in, out, s); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func Convert_core_PodSecurityContext_To_v1_PodSecurityContext(in *core.PodSecurityContext, out *v1.PodSecurityContext, s conversion.Scope) error {
+	if err := autoConvert_core_PodSecurityContext_To_v1_PodSecurityContext(in, out, s); err != nil {
+		return err
+	}
+
+	return nil
+}
+
```
why was this needed?

pjbgf replied:
This was automatically generated and I wasn't sure whether it was needed. Now it is removed.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3577,6 +3599,16 @@ func ValidateSeccompPodAnnotations(annotations map[string]string, fldPath *field
+	case core.SeccompProfileLocalhost, core.SeccompProfileTypeDefault, core.SeccompProfileUnconfined:
+		return nil
+	default:
+		return field.NotSupported(fldPath, seccompProfileType, []string{string(core.SeccompProfileUnconfined), string(core.SeccompProfileUnconfined)})
```
this should enumerate the three valid values

pjbgf replied:
Fixed now.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3550,15 +3550,37 @@ func validatePodAffinity(podAffinity *core.PodAffinity, fldPath *field.Path) fie
+	}
+
+	if sp.Type != core.SeccompProfileLocalhost && sp.LocalhostProfile != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, sp, "profileLocalhost can only be set when seccomp type is Localhost"))
```
```suggestion
		allErrs = append(allErrs, field.Invalid(fldPath.Child("localhostProfile"), sp, "can only be set when seccomp type is Localhost"))
```

pjbgf replied:
Just updated accordingly.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3550,15 +3550,37 @@ func validatePodAffinity(podAffinity *core.PodAffinity, fldPath *field.Path) fie
+	}
+
+	if sp.Type == core.SeccompProfileLocalhost && sp.LocalhostProfile == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, sp, "profileLocalhost must be set when seccomp type is Localhost"))
```
```suggestion
		allErrs = append(allErrs, field.Required(fldPath.Child("localhostProfile"), "must be set when seccomp type is Localhost"))
```

pjbgf replied:
Just updated accordingly.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3550,15 +3550,37 @@ func validatePodAffinity(podAffinity *core.PodAffinity, fldPath *field.Path) fie
+	if err := ValidateSeccompProfileType(fldPath.Child("type"), sp.Type); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
```
restructure to avoid repeated checks of the same conditions:

```go
if sp.Type == core.SeccompProfileLocalhost {
  if sp.LocalhostProfile == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("localhostProfile"), ...
  } else {
    allErrs = append(allErrs, validateLocalDescendingPath(...
  }
} else {
  if sp.LocalhostProfile != nil {
    allErrs = append(allErrs, field.Invalid(fldPath.Child("localhostProfile"), ...
  }
}
```

pjbgf replied:
Amended the code with your suggestion.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3550,15 +3550,37 @@ func validatePodAffinity(podAffinity *core.PodAffinity, fldPath *field.Path) fie
+	}
+
+	if sp.Type == core.SeccompProfileLocalhost && sp.LocalhostProfile != nil {
+		allErrs = append(allErrs, validateLocalDescendingPath(*sp.LocalhostProfile, fldPath)...)
```
```suggestion
		allErrs = append(allErrs, validateLocalDescendingPath(*sp.LocalhostProfile, fldPath.Child("localhostProfile"))...)
```

liggitt commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -202,8 +202,37 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus) *kubecontainer.Runtim
+			return v1.SeccompLocalhostProfileNamePrefix + fname
+		}
+
+		return ""
```
this path shouldn't be possible, right? what should we do if we get here? what does returning "" do?

pjbgf replied:
Returning "" causes dockershim to use the unconfined profile:
https://github.com/kubernetes/kubernetes/blob/a79c711191d5c0a9dca4fbaba26ef35e476f5871/pkg/kubelet/dockershim/helpers_linux.go#L52

I removed this and made the behaviour a bit more explicit.

liggitt commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -202,8 +202,37 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus) *kubecontainer.Runtim
 // getSeccompProfileFromAnnotations gets seccomp profile from annotations.
 // It gets pod's profile if containerName is empty.
+// Seccomp Annotations is deprecated and should be replaced by getSeccompProfile.
 func (m *kubeGenericRuntimeManager) getSeccompProfileFromAnnotations(annotations map[string]string, containerName string) string {
```
I would suggest a single method that looks at container/pod fields and annotations, otherwise this is likely to attract call sites that don't honor the fields

pjbgf replied:
Done, I removed the older method.

liggitt commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -202,8 +202,37 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus) *kubecontainer.Runtim
+		return fieldProfile(podSecContext.SeccompProfile)
+	}
+
+	return m.getSeccompProfileFromAnnotations(annotations, containerName)
```
the KEP described a lookup path that did the following:

1. container-specific field
2. container-specific annotation
3. pod-level field
4. pod-level annotation

This implementation would honor a pod-level field over a container-level annotation

pjbgf replied:
I added an unit test to cover the missing scenario and corrected the priority order.

liggitt commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -144,7 +144,7 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxLinuxConfig(pod *v1.Pod) (
 		SecurityContext: &runtimeapi.LinuxSandboxSecurityContext{
 			Privileged:         kubecontainer.HasPrivilegedContainer(pod),
-			SeccompProfilePath: m.getSeccompProfileFromAnnotations(pod.Annotations, ""),
+			SeccompProfilePath: m.getSeccompProfile(pod.Annotations, "", nil, nil),
```
why are we not passing in the security context fields here?

pjbgf replied:
Well spotted. I am now only setting the PodSecurityContext, as that is analogous to the existing behaviour.

pjbgf replied:
I created some tests around this to check that container-level seccomp field and annotation are ignored. 

Currently, there is no way to set seccomp to a sandbox container at container-level. Am I right to assume we are keeping this behaviour? @tallclair @liggitt

liggitt commented:
I expected this PR to contain the validation bits that ensured if annotations and fields were both present on create, they matched. That logically belongs with the field addition / validation bit, even if a follow up will automatically do some translation between the two.

saschagrunert commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -539,3 +546,82 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
+	return nil
+}
+
+func syncSeccompAnnotationAndFieldForPod(annotationKey, annotationValue string, field *core.SeccompProfile, pod *v1.Pod) (*v1.SeccompProfile, error) {
```
Let's rename this function and change its signature:
```suggestion
func validateSeccompAnnotationAndProfile(annotationKey, annotationValue string, field *core.SeccompProfile, pod *v1.Pod) (error) {
```
This way I can add a new function later on without having the need for changing this one again.

pjbgf replied:
That makes sense, just amended that. :+1:

saschagrunert commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -539,3 +546,82 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
+	return nil
+}
+
+func validateSeccompAnnotationAndProfile(annotationKey, annotationValue string, field *core.SeccompProfile, pod *v1.Pod) (*v1.SeccompProfile, error) {
```
Let's remove the unused return parameter too:
```suggestion
func validateSeccompAnnotationAndProfile(annotationKey, annotationValue string, field *core.SeccompProfile, pod *v1.Pod) error {
```

pjbgf replied:
Well spotted. I removed that and updated the surrounding code.

saschagrunert commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -539,3 +546,82 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
+		podInField = in.Spec.SecurityContext.SeccompProfile
+	}
+
+	err := validateSeccompAnnotationAndProfile(
+		v1.SeccompPodAnnotationKey, annotationValue, podInField, out,
+	)
+	if err != nil {
+		return errors.Wrapf(err, "synchronizing annotations and security context field for pod: %+v", in)
+	}
```
nit
```suggestion
	if err := validateSeccompAnnotationAndProfile(
		v1.SeccompPodAnnotationKey, annotationValue, podInField, out,
	); err != nil {
		return errors.Wrapf(err, "synchronizing annotations and security context field for pod: %+v", in)
	}
```

saschagrunert commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -539,3 +546,82 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
+			}
+		}
+
+		err := validateSeccompAnnotationAndProfile(
+			annotationKey, annotationValue, ctrInField, out,
+		)
+		if err != nil {
+			return errors.Wrapf(err, "synchronizing annotations and security context field for container: %+v", ctr)
+		}
```
nit
```suggestion
		if err := validateSeccompAnnotationAndProfile(
			annotationKey, annotationValue, ctrInField, out,
		); err != nil {
			return errors.Wrapf(err, "synchronizing annotations and security context field for container: %+v", ctr)
		}
```

saschagrunert commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -539,3 +546,82 @@ func Convert_core_PersistentVolumeSpec_To_v1_PersistentVolumeSpec(in *core.Persi
+func validateSeccompAnnotationAndProfile(annotationKey, annotationValue string, field *core.SeccompProfile, pod *v1.Pod) error {
+	// If both seccomp annotations and fields are specified, the values must
+	// match.
+	if field != nil && annotationValue != "" {
```
nit, to early return
```suggestion
	if field == nil || annotationValue == "" {
		return nil
	}
```

saschagrunert reviewed (approved):
/lgtm

Just some nits which are non-blocking from my side.

pjbgf commented:
@liggitt thank you for all the feedback. I worked through them, PTAL

pjbgf commented:
/retest

pjbgf commented:
/retest

pjbgf commented:
Test failure does not seem to be related to the changes on the PR, so will retry once more:
```
[sig-storage] PersistentVolumes NFS with multiple PVs and PVCs all in same ns should create 2 PVs and 4 PVCs: test write access expand_less | 5m36s
-- | --
test/e2e/storage/persistent_volumes.go:233 May 27 20:16:13.671: Unexpected error:     <*errors.errorString \| 0xc0015eb9a0>: {         s: "pod \"pvc-tester-2h4b4\" did not exit with Success: pod \"pvc-tester-2h4b4\" failed to reach Success: Gave up after waiting 5m0s for pod \"pvc-tester-2h4b4\" to be \"Succeeded or Failed\"",     }     pod "pvc-tester-2h4b4" did not exit with Success: pod "pvc-tester-2h4b4" failed to reach Success: Gave up after waiting 5m0s for pod "pvc-tester-2h4b4" to be "Succeeded or Failed" occurred test/e2e/storage/persistent_volumes.go:238
```

/retest

saschagrunert commented:
I agree,
/test pull-kubernetes-e2e-kind

liggitt commented on pkg/apis/core/v1/conversion.go:
```diff
@@ -382,6 +385,10 @@ func Convert_core_Pod_To_v1_Pod(in *core.Pod, out *v1.Pod, s conversion.Scope) e
 	// remove this once the oldest supported kubelet no longer honors the annotations over the field.
 	out.Annotations = dropInitContainerAnnotations(out.Annotations)
 
+	if err := HandleSeccompVersionSkew(in, out); err != nil {
+		return errors.Wrap(err, "handling seccomp version skew")
+	}
```
Conversion is not the right place for this... an error here means the pod won't even reach validation code in the API server.

podStrategy#PrepareForCreate and podStrategy#PrepareForUpdate are where I would expect us to make these changes

pjbgf replied:
Thank you for pointing me to the right direction, I wasn't quite sure where in the code that was to take place. PATL

liggitt replied:
I still don't see anything in podStrategy#PrepareForCreate or podStrategy#PrepareForUpdate that does the annotation<->field population

pjbgf commented:
/test pull-kubernetes-e2e-kind-ipv6

saschagrunert reviewed (approved):
/lgtm

kow3ns commented:
/remove-sig apps

saschagrunert commented:
/retest

liggitt commented on staging/src/k8s.io/api/core/v1/annotation_key_constants.go:
```diff
@@ -39,15 +39,24 @@ const (
 	SeccompContainerAnnotationKeyPrefix string = "container.seccomp.security.alpha.kubernetes.io/"
 
 	// SeccompProfileRuntimeDefault represents the default seccomp profile used by container runtime.
+	// Seccomp annotations are now deprecated. This should be replaced by SeccompProfileTypeRuntimeDefault.
```
```suggestion
	// Deprecated: set a pod or container security context `seccompProfile` of type "RuntimeDefault" instead
```

liggitt commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -94,6 +94,8 @@ func (podStrategy) Validate(ctx context.Context, obj runtime.Object) field.Error
 	}
 	allErrs := validation.ValidatePodCreate(pod, opts)
 	allErrs = append(allErrs, validation.ValidateConditionalPod(pod, nil, field.NewPath(""))...)
+	allErrs = append(allErrs, validation.ValidateSeccompAnnotationsAndFields(pod, field.NewPath(""))...)
```
is there a reason not to call this from ValidatePodCreate (and ValidatePodUpdate), rather than requiring callers to add additional validation calls to get the entire pod validated?

liggitt commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -202,31 +202,51 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus) *kubecontainer.Runtim
+		if scmp.Type == v1.SeccompProfileTypeRuntimeDefault {
+			return v1.SeccompProfileRuntimeDefault
+		}
+		if scmp.Type == v1.SeccompProfileTypeLocalhost && scmp.LocalhostProfile != nil {
+			fname := filepath.Join(m.seccompProfileRoot, *scmp.LocalhostProfile)
+			return v1.SeccompLocalhostProfileNamePrefix + fname
+		}
```
if `LocalhostProfile == nil` here, should this be treated like an error condition and return "", rather than fall through and return SeccompProfileNameUnconfined?

pjbgf replied:
that makes sense, I update it accordingly

liggitt commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -202,31 +202,51 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus) *kubecontainer.Runtim
-	profile, profileOK := annotations[v1.SeccompPodAnnotationKey]
+func (m *kubeGenericRuntimeManager) getSeccompProfile(annotations map[string]string, containerName string,
+	podSecContext *v1.PodSecurityContext, containerSecContext *v1.SecurityContext) string {
+	fieldProfile := func(scmp *v1.SeccompProfile) string {
```
move fieldProfile and annotationProfile to be standalone functions (can pass in the profileroot path as a param)... it'll make unit tests easier and we don't need to define them as local functions

pjbgf replied:
extracted the functions and added a new set of tests around it.

liggitt commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -202,31 +202,51 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus) *kubecontainer.Runtim
 	}
 
-	return profile
+	return v1.SeccompProfileNameUnconfined
```
should this return "" to match previous behavior?

liggitt commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -202,31 +202,51 @@ func toKubeRuntimeStatus(status *runtimeapi.RuntimeStatus) *kubecontainer.Runtim
+			fname := filepath.Join(m.seccompProfileRoot, *scmp.LocalhostProfile)
+			return v1.SeccompLocalhostProfileNamePrefix + fname
+		}
+		return v1.SeccompProfileNameUnconfined
```
I'd prefer an exact match here rather than defaulting an unknown type to "unconfined":
```go
if scmp.Type == v1.SeccompProfileTypeUnconfined {
  return v1.SeccompProfileNameUnconfined
}
return ""
```

pjbgf replied:
Now it is all aligned with existing behaviour - non-expected state returns "".

pjbgf committed:
Add seccomp enforcement and validation based on new GA fields

Adds seccomp validation.

This ensures that field and annotation values must match when present.

Co-authored-by: Sascha Grunert <sgrunert@suse.com>

pjbgf commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3577,6 +3599,18 @@ func ValidateSeccompPodAnnotations(annotations map[string]string, fldPath *field
+var supportedSeccompProfileTypes = sets.NewString(string(core.SeccompProfileTypeLocalhost), string(core.SeccompProfileTypeRuntimeDefault), string(core.SeccompProfileTypeUnconfined))
+
+// ValidateSeccompProfileType tests that the argument is a valid SeccompProfileType.
+func ValidateSeccompProfileType(fldPath *field.Path, seccompProfileType core.SeccompProfileType) *field.Error {
```
Agreed and changed

pjbgf commented on cluster/images/conformance/go-runner/BUILD:
```diff
@@ -41,6 +41,7 @@ go_test(
         "env_test.go",
         "tar_test.go",
     ],
+    data = glob(["testdata/**"]),
```
This keeps being added automatically on `make update`, I have reverted it now.

pjbgf commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3724,6 +3762,61 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+		}
+
+	case core.SeccompProfileTypeLocalhost:
+		if strings.TrimPrefix(annotationValue, v1.SeccompLocalhostProfileNamePrefix) != *seccompField.LocalhostProfile {
```
> also need to make sure that annotationValue has a prefix of `v1.SeccompLocalhostProfileNamePrefix` and fail if it doesn't
This validation currently exists here: https://github.com/kubernetes/kubernetes/blob/49f6a5f9cd3d1b5c1a93baaafeafad8edf3e603b/pkg/apis/core/validation/validation.go#L3593

Do we need to do that here too?

pjbgf commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3226,8 +3226,38 @@ type PodSecurityContext struct {
+// Only one profile source may be set.
+// +union
+type SeccompProfile struct {
+	// +unionDiscriminator
```
add a bit more context here

pjbgf commented:
/retest

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3756,69 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+}
+
+// ValidateSeccompAnnotationsAndFields iterates through all containers and ensure that when both seccompProfile and seccomp annotations exist they match.
+func ValidateSeccompAnnotationsAndFields(objectMeta metav1.ObjectMeta, podSpec *core.PodSpec, fldPath *field.Path) field.ErrorList {
```
unexport unless used outside this package:
```suggestion
func validateSeccompAnnotationsAndFields(objectMeta metav1.ObjectMeta, podSpec *core.PodSpec, fldPath *field.Path) field.ErrorList {
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3756,69 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
 	if len(pod.Spec.EphemeralContainers) > 0 {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("ephemeralContainers"), "cannot be set on create"))
 	}
+	allErrs = append(allErrs, ValidateSeccompAnnotationsAndFields(pod.ObjectMeta, &pod.Spec, field.NewPath(""))...)
```
this should also be called from ValidatePodTemplateSpec so we validate annotation/podSpec combinations in workload objects (add a unit test for that)

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3756,69 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+		}
+	}
+
+	podshelper.VisitContainersWithPath(podSpec, func(c *core.Container, fldPath *field.Path) bool {
```
interestingly, the fldPath passed into ValidateSeccompAnnotationsAndFields is lost here... that becomes important when we're validating a pod spec template inside an object like a deployment... the field path to an error there should be something like `spec.template.spec.securityContext.seccompProfile`, etc

liggitt replied:
looks like we need a test that exercises initContainers as well to make sure the field path is correct

pjbgf replied:
added this unit test to assert all container types:

```
			"all container types with template pod path",
			field.NewPath("template", "spec"),
			&api.PodSpec{
				Containers: []api.Container{
					{Name: "c1"},
					{Name: "c2"},
				},
				InitContainers: []api.Container{
					{Name: "i1"},
					{Name: "i2"},
				},
				EphemeralContainers: []api.EphemeralContainer{
					{EphemeralContainerCommon: api.EphemeralContainerCommon{Name: "e1"}},
					{EphemeralContainerCommon: api.EphemeralContainerCommon{Name: "e2"}},
				},
			},
			[]string{"template.spec.initContainers[0]", "template.spec.initContainers[1]", "template.spec.containers[0]", "template.spec.containers[1]", "template.spec.ephemeralContainers[0]", "template.spec.ephemeralContainers[1]"},
```

pjbgf reviewed (commented):
@liggitt Just finished with the last feedback, PTAL.

pjbgf commented:
/retest

pjbgf commented:
/retest

saschagrunert commented:
/retest

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,75 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
 }
 
+// ValidateSeccompAnnotationsAndFields iterates through all containers and ensure that when both seccompProfile and seccomp annotations exist they match.
+func validateSeccompAnnotationsAndFields(objectMeta metav1.ObjectMeta, podSpec *core.PodSpec, fldPath *field.Path) field.ErrorList {
```
for clarity:
```suggestion
func validateSeccompAnnotationsAndFields(objectMeta metav1.ObjectMeta, podSpec *core.PodSpec, specPath *field.Path) field.ErrorList {
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,75 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.SeccompProfile != nil {
+		// If both seccomp annotations and fields are specified, the values must match.
+		if annotation, found := objectMeta.Annotations[v1.SeccompPodAnnotationKey]; found {
+			fldPath = fldPath.Child("securityContext").Child("seccompProfile")
+			allErrs = append(allErrs, validateSeccompAnnotationsAndFieldsMatch(annotation, podSpec.SecurityContext.SeccompProfile, fldPath))
```
this overrides fldPath which may be used later. suggest:
```suggestion
			seccompPath := specPath.Child("securityContext").Child("seccompProfile")
			allErrs = append(allErrs, validateSeccompAnnotationsAndFieldsMatch(annotation, podSpec.SecurityContext.SeccompProfile, seccompPath))
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,75 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+	}
+
+	index := 0
+	fldPath = fldPath.Child("containers")
```
VisitContainersWithPath visits containers of all types (initContainers, containers, ephemeralContainers), so a single field path and increasing index is not correct... 

I would suggest the following:
1. change the signature of VisitContainersWithPath to `VisitContainersWithPath(podSpec *api.PodSpec, specPath *field.Path, visitor ContainerVisitorWithPath)`
2. Update all other existing call sites to pass in `field.NewPath("spec")` as the spec path
3. Update this call site to pass in the `specPath` arg

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,75 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+
+		key := v1.SeccompContainerAnnotationKeyPrefix + c.Name
+		if annotation, found := objectMeta.Annotations[key]; found {
+			allErrs = append(allErrs, validateSeccompAnnotationsAndFieldsMatch(annotation, field,
+				fldPath.Index(index).Child("securityContext").Child("seccompProfile")))
```
```suggestion
			allErrs = append(allErrs, validateSeccompAnnotationsAndFieldsMatch(annotation, field,
				cFldPath.Child("securityContext").Child("seccompProfile")))
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,75 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
 	if len(pod.Spec.EphemeralContainers) > 0 {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("ephemeralContainers"), "cannot be set on create"))
 	}
+	allErrs = append(allErrs, validateSeccompAnnotationsAndFields(pod.ObjectMeta, &pod.Spec, field.NewPath(""))...)
```
```suggestion
	allErrs = append(allErrs, validateSeccompAnnotationsAndFields(pod.ObjectMeta, &pod.Spec, field.NewPath("spec"))...)
```

liggitt commented:
Once the validation changes are done, the API and validation bits lgtm. We will hold this until the annotation<->field setting bits on pod creation are complete in https://github.com/kubernetes/kubernetes/pull/91408 before merging this

pjbgf commented:
/retest

liggitt commented:
API change and validation lgtm

/hold pending https://github.com/kubernetes/kubernetes/pull/91408 being ready

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,71 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+		}
+
+	case core.SeccompProfileTypeRuntimeDefault:
+		if annotationValue != v1.SeccompProfileRuntimeDefault {
```
should this also allow DeprecatedSeccompProfileDockerDefault?

pjbgf replied:
Yep, that makes a lot of sense. I will be pushing a fix for this imminently.

liggitt commented:
noticed one validation gap in https://github.com/kubernetes/kubernetes/pull/91381/files#r447910309

pjbgf commented on pkg/apis/core/pods/helpers_test.go:
```diff
@@ -105,11 +112,30 @@ func TestVisitContainersWithPath(t *testing.T) {
 			},
 			[]string{"spec.initContainers[0]", "spec.initContainers[1]", "spec.containers[0]", "spec.containers[1]", "spec.ephemeralContainers[0]", "spec.ephemeralContainers[1]"},
 		},
+		{
+			"all container types with template pod path",
+			field.NewPath("template", "spec"),
+			&api.PodSpec{
+				Containers: []api.Container{
+					{Name: "c1"},
+					{Name: "c2"},
+				},
+				InitContainers: []api.Container{
+					{Name: "i1"},
+					{Name: "i2"},
+				},
+				EphemeralContainers: []api.EphemeralContainer{
+					{EphemeralContainerCommon: api.EphemeralContainerCommon{Name: "e1"}},
+					{EphemeralContainerCommon: api.EphemeralContainerCommon{Name: "e2"}},
+				},
+			},
+			[]string{"template.spec.initContainers[0]", "template.spec.initContainers[1]", "template.spec.containers[0]", "template.spec.containers[1]", "template.spec.ephemeralContainers[0]", "template.spec.ephemeralContainers[1]"},
```
@liggitt added this test to cover the change on `VisitContainersWithPath`.

pjbgf commented:
/retest

pjbgf commented:
/retest

pjbgf commented:
/retest

saschagrunert commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,71 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
 	if len(pod.Spec.EphemeralContainers) > 0 {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("ephemeralContainers"), "cannot be set on create"))
 	}
+	allErrs = append(allErrs, validateSeccompAnnotationsAndFields(pod.ObjectMeta, &pod.Spec, field.NewPath("spec"))...)
```
I'm not sure if this is the only place, but there might be the case that we end-up having having an error list of `[<nil>]` here. When I'm trying out the PR https://github.com/kubernetes/kubernetes/pull/91408, then the issue is already reproducible when running hack/local-up-cluster.sh (see the kubedns replicaset).

pjbgf replied:
Thank you for pointing it out, I have just pushed a fix for it.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,71 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+		// If both seccomp annotations and fields are specified, the values must match.
+		if annotation, found := objectMeta.Annotations[v1.SeccompPodAnnotationKey]; found {
+			seccompPath := specPath.Child("securityContext").Child("seccompProfile")
+			allErrs = append(allErrs, validateSeccompAnnotationsAndFieldsMatch(annotation, podSpec.SecurityContext.SeccompProfile, seccompPath))
```
this can append a nil error, which we don't want

liggitt replied:
adda unit test that would have caught this

pjbgf replied:
added validation and unit tests for this scenario.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3720,10 +3758,71 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+		key := v1.SeccompContainerAnnotationKeyPrefix + c.Name
+		if annotation, found := objectMeta.Annotations[key]; found {
+			seccompPath := cFldPath.Child("securityContext").Child("seccompProfile")
+			allErrs = append(allErrs, validateSeccompAnnotationsAndFieldsMatch(annotation, field, seccompPath))
```
this can append a nil error, which we don't want

liggitt replied:
add a unit test that would have caught this

pjbgf replied:
added validation and unit tests for this scenario too

saschagrunert commented:
/test pull-kubernetes-kubemark-e2e-gce-big

liggitt commented:
/lgtm
/approve
/priority important-soon

liggitt commented:
/hold cancel
https://github.com/kubernetes/kubernetes/pull/91408 has lgtm
